pyomo/core/tests/unit/test_block.py

```diff
@@ -647,6 +647,102 @@ def test_set_attr(self):
         self.block.x = None
         self.assertEqual(self.block.x._value, None)
 
+        ### creation of a circular reference
+        b = Block(concrete=True)
+        b.c = Block()
+        with self.assertRaisesRegexp(
+                ValueError, "Cannot assign the top-level block as a subblock "
+                "of one of its children \(c\): creates a circular hierarchy"):
+            b.c.d = b
+
+    def test_set_value(self):
+        b = Block(concrete=True)
+        with self.assertRaisesRegexp(
+                RuntimeError, "Block components do not support assignment "
+                "or set_value"):
+            b.set_value(None)
+
+        b.b = Block()
+        with self.assertRaisesRegexp(
+                RuntimeError, "Block components do not support assignment "
+                "or set_value"):
+            b.b = 5
+
+    def test_transfer_attributes_from(self):
+        b = Block(concrete=True)
+        b.x = Var()
+        b.y = Var()
+        c = Block(concrete=True)
+        c.z = Param(initialize=5)
+        c.x = c_x = Param(initialize=5)
+        c.y = c_y = 5
+
+        b.transfer_attributes_from(c)
+        self.assertEqual(list(b.component_map()), ['z','x'])
+        self.assertEqual(list(c.component_map()), [])
+        self.assertIs(b.x, c_x)
+        self.assertIs(b.y, c_y)
+
+        b = Block(concrete=True)
+        b.x = Var()
+        b.y = b_y = Var()
+        c = Block(concrete=True)
+        c.z = Param(initialize=5)
+        c.x = c_x = Param(initialize=5)
+        c.y = c_y = 5
+
+        b.transfer_attributes_from(c, guarantee_components={'y','x'})
```

Comment:
Identifying the components in `guarantee_components` by their local names makes me a little nervous. Will this approach break down if you want to "guarantee" components deeper in the hierarchy when transferring attributes from nested blocks?

Reply:
Because this method is not recursive, and only moves/manipulates components directly attached to this block (and the source block), I believe that local names are perfectly safe.